### ISSUE TITLE
`ansi` command should not output ANSI color codes when `use_ansi_coloring` is disabled in config

### ISSUE THREAD

**Issue by NotTheDr01ds** [nushell/nushell#14194]
### Describe the bug

This is the root cause of #14043.

In (unrelated to 14043) digging around in the `ansi` command, it comes to mind that the reason why user-land ANSI was still coloring output in #14043 is that the `ansi` command itself still outputs color codes, even when `use_ansi_coloring` is false.

This shouldn't happen.

The `ansi` command actually checks the `use_ansi_coloring` option, but only to determine if color previews should be shown with `ansi --list/-l`.  It should also check before outputting a color (or attribute) code.

### How to reproduce

Note: to reproduce correctly, each line of the following must be executed separately; otherwise the environment won't be re-read before the command is executed ...

```nu
$env.use_ansi_coloring = false
$"(ansi red)This shouldn't be Red, but it is(ansi reset)"
```

### Expected behavior

I expect the `ansi` command to disregard color and attribute codes when `use_ansi_coloring` is `false`.  ANSI codes related to cursor positioning, clearing the screen, and others *should* still be output.

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.98.1                                   |
| major              | 0                                        |
| minor              | 98                                       |
| patch              | 1                                        |
| branch             | virtual-mods                             |
| commit_hash        | 1dc58b84a85a7d7e8fa4a9fec866582eddcfc036 |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.79.0 (129f3b996 2024-06-10)      |
| rust_channel       | 1.79.0-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.79.0 (ffa9cf99a 2024-06-03)      |
| build_time         | 2024-10-09 13:39:29 -04:00               |
| build_rust_channel | release                                  |
| allocator          | mimalloc                                 |
| features           | default, sqlite, trash                   |
| installed_plugins  |                                          |

**Comment by weirdan** [comment]
I would expect it to work even if `use_ansi_coloring` is `false`. IMHO the command is too low-level to take into account shell preferences.

**Comment by NotTheDr01ds** [comment]
I can probably fix this:

* (More Easily) By just grouping color and attribute codes in the first x elements in the codes array.  This is already how `ansi -l` determines which codes to preview (or not) based on `use_ansi_coloring`
* (More Safely) By adding a new boolean field to the `AnsiCode` struct indicating whether or not that code should be output when ANSI coloring is disabled.

The problem with the first method (as well as for `ansi -l` is that it relies on every subsequent contributor understanding where to place new items in the vector.  This can be handled with comments, but it's not foolproof, of course.

**Comment by NotTheDr01ds** [comment]
@weirdan The problem is that any Nushell code that uses it (e.g., `std` and others) doesn't honor the setting without extensive additional logic.  I expect the `ansi` command to have that knowledge built-in, personally.

**Comment by NotTheDr01ds** [comment]
We could have a flag to override the `use_ansi_coloring` setting if needed.

**Comment by weirdan** [comment]
The whole reason people have `ansi red` in their scripts is because they lack a more high-level abstraction like `show error ...` (that could take `use_ansi_coloring` into account).

**Comment by NotTheDr01ds** [comment]
@weirdan Can you elaborate?  Are you saying that someone would turn off `use_ansi_coloring` and then put a `ansi red` in their script because `error make` then doesn't appear in color?

I think:

1. It's unlikely that the same person would turn off `use_ansi_coloring` and then turn around and put an `ansi red` (etc.) in a script.
2. In the event the script author *wanted* colors while debugging, wouldn't they just set `use_ansi_coloring: true`?
3. It's far more likely that a script author has put `ansi red` in a script and then a *user* of that script/module/etc. wants to disable it via `use_ansi_coloring: false`.
4. In the rare event that (2) above doesn't work for a use-case, we could have a `--force` flag on `ansi`, I guess.  But I'd hate for script authors to force it on *other* users who have disabled it.

Ultimately the most likely use case for `$env.config.use_ansi_coloring: false` is probably *accessibility*.  Color-blind individuals may have a difficult time differentiating certain colors from the background, so they turn off all coloring.  We need to support that even for third-party (and first party, as seen in `std/banner` and the prompts), I believe.

**Comment by weirdan** [comment]
> Are you saying that someone would turn off use_ansi_coloring and then put an ansi red in their script because error make then doesn't appear in color?

No, I'm saying that most users (like 95% of them) shouldn't need `ansi red` at all. What they need is something like `show message --style=error ...` instead. Similar to `std log error` but less 'loggy'. And whatever implements that (let's call it a widget library for the sake of argument) should account for `use_ansi_coloring` value.

**Comment by NotTheDr01ds** [comment]
Alternatively, I could see just recommending a `hook.display_output` where everything is `ansi strip`'d.  Of course, then I'm not quite sure what the use of `$env.config.use_ansi_coloring` would really be.

**Comment by weirdan** [comment]
>  If we fix the problem in the internal ansi statement, then we fix it for all third-party modules/commands/etc.

Because the problem is not in the internal `ansi` command. It does what it says - generates an ANSI escape sequence for the specified colour. I'd say it would have a problem if it stopped doing that in some circumstances. It's bad enough with terminal colour schemes that also remap colours.

**Comment by NotTheDr01ds** [comment]
But the whole point of `use_ansi_coloring: false` is to tell internal Nushell commands to *not* generate ANSI escape sequences for colors. IMHO we have a problem that the `ansi` command is not respecting that.  Either way, someone is going to considering it a problem, apparently ;-).

**Comment by cptpiepmatz** [comment]
Maybe relevant #14647

**Comment by cptpiepmatz** [comment]
I totally understand @weirdan's point but I think this makes writing scripts hard. Just throwing some colors into your script is pretty easy and most often enough. Having wrappers around some colors is great but hard to catch every use case. 

Also with #14647 landed, the behavior of `$env.config.use_ansi_coloring` is way more complex. I think that value should only ever be set by nushell code and never read as internally it also respects some env variables.

Having `ansi` respect that config makes it easy to write scripts with colors and also fixes all the standard library code that uses that.  If someone wants to enforce colors, they can store the value in `$env.config.use_ansi_coloring`, set it to `true`, do their coloring and later reset it.

@NotTheDr01ds @132ikl your thoughts?

**Comment by 132ikl** [comment]
We talked about this a bit more in the Discord and one of the meetings a while back, and after those discussions I think that the best solution would be to have better mechanisms for working with color in Nushell, and having those respect the envvars (which is kind of along the lines that @weirdan was suggesting). I think styled strings would still be a good option, but there hasn't really been much discussion about what that might look like yet. The general idea is that you would associate styles with spans of strings, and then colorize them though a mechanism a la the `table` command. I would be interested in discussion this a bit further, maybe we should start up a drawing board thread in Discord.

Barring a higher-level mechanism for coloring strings, I'm not really sure what to do with this. I can definitely get behind the argument that `ansi` should always just emit ansi escape characters, but the accessibility concern is also very real. I do think at the very least we should have some interim mechanism for disabling ANSI coloring altogether, whether this is `NO_COLOR` or something else.

**Comment by cptpiepmatz** [comment]
I see now, yeah, I'm gonna open up a drawing board for further discussions.
